crates/cargo-bsp/src/cargo_communication/check/cargo_message_to_package_info.rs

```diff
@@ -0,0 +1,136 @@
+//! Maps information from Cargo Messages produces by `cargo check` command to
+//! RustPackage information.
+
+use crate::utils::uri::file_uri;
+use bsp_types::extensions::RustCfgOptions;
+use bsp_types::Uri;
+use cargo_metadata::{Artifact, BuildScript, Package};
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+
+const DYNAMIC_LIBRARY_EXTENSIONS: [&str; 3] = ["dll", "so", "dylib"];
+const PROC_MACRO: &str = "proc-macro";
+
+#[derive(Default)]
+struct SplitVersion {
+    major: String,
+    minor: String,
+    pre_release: String,
+    patch: String,
+}
+
+pub(super) fn map_cfg_options(script: Option<&BuildScript>) -> Option<RustCfgOptions> {
+    script.map(|s| {
+        let mut key_value_options: HashMap<String, Vec<String>> = HashMap::new();
+        let mut name_options = Vec::new();
+
+        s.cfgs.iter().for_each(|cfg| {
+            let mut parts = cfg.splitn(2, '=');
+            let key = parts.next();
+            let value = parts.next().map(|v| v.trim_matches('"').to_string());
+
+            if let Some(k) = key {
+                if let Some(v) = value {
+                    if let Entry::Vacant(e) = key_value_options.entry(k.to_string()) {
+                        e.insert(vec![v]);
+                    } else {
+                        key_value_options.get_mut(k).unwrap().push(v);
+                    }
+                } else {
+                    name_options.push(k.to_string());
+                }
+            }
+        });
+
+        RustCfgOptions {
+            key_value_options,
+            name_options,
+        }
+    })
+}
+
+fn split_version(version: String) -> SplitVersion {
+    if let Some((major, rest)) = version.split_once('.') {
+        if let Some((minor, rest)) = rest.split_once('.') {
+            let (patch, pre_release) = rest
+                .split_once('-')
+                .map(|(s1, s2)| (s1.to_string(), s2.to_string()))
+                .unwrap_or((rest.to_string(), String::default()));
+            return SplitVersion {
+                major: major.to_string(),
+                minor: minor.to_string(),
+                pre_release,
+                patch,
+            };
+        }
+    }
+    SplitVersion::default()
```

Comment:
default to będzie wszystko puste, czy to ma sens? nie lepiej zwrócić Option?
also, czy nie ma jakiejś biblioteki do parsowania tego? https://docs.rs/semver/latest/semver/

Reply:
ta biblioteka to wręcz jest "cargo flavor"

Reply:
Jeśli chodzi o ten default, to starałam się to dopasowywać do tego, czego oczekiwałby od nas intellij-rust i jak on to rozwiązuje . W intellij-rust, jeżeli nie można podzielić wersji, to jest dawany pusty string pod każdą z wartości

Reply:
Chyba jak to pisałam to zawiesił mi się mózg, bo nie zauważyłam, że mam potrzebne informacje na wyciągnięcie ręki. Więc już poprawiłam